File: modules/angular2/src/facade/collection.dart
library facade.collection;

import 'dart:collection' show IterableBase, Iterator;
export 'dart:core' show Map, List, Set;
import 'dart:math' show max, min;

class MapIterator extends Iterator<List> {
  final Iterator _iterator;
  final Map _map;

  MapIterator(Map map)
      : _map = map,
        _iterator = map.keys.iterator;

  bool moveNext() => _iterator.moveNext();

  List get current {
    return _iterator.current != null
        ? [_iterator.current, _map[_iterator.current]]
        : null;
  }
}

class IterableMap extends IterableBase<List> {
  final Map _map;

  IterableMap(Map map) : _map = map;

  Iterator<List> get iterator => new MapIterator(_map);
}

class MapWrapper {
  static Map create() => {};
  static Map clone(Map m) => new Map.from(m);
  static Map createFromStringMap(Map m) => m;
  static Map createFromPairs(List pairs) => pairs.fold({}, (m, p) {
    m[p[0]] = p[1];
    return m;
  });
  static get(Map m, k) => m[k];
  static void set(Map m, k, v) {
    m[k] = v;
  }
  static contains(Map m, k) => m.containsKey(k);
  static forEach(Map m, fn(v, k)) {
    m.forEach((k, v) => fn(v, k));
  }
  static int size(Map m) => m.length;
  static void delete(Map m, k) {
    m.remove(k);
  }
  static void clear(Map m) {
    m.clear();
  }
  static void clearValues(Map m) {
    for (var k in m.keys) {
      m[k] = null;
    }
  }
  static Iterable iterable(Map m) => new IterableMap(m);
  static Iterable keys(Map m) => m.keys;
  static Iterable values(Map m) => m.values;
}

class StringMapWrapper {
  static Map create() => {};
  static bool contains(Map map, key) => map.containsKey(key);
  static get(Map map, key) => map[key];
  static void set(Map map, key, value) {
    map[key] = value;
  }
  static void delete(Map m, k) {
    m.remove(k);
  }
  static void forEach(Map m, fn(v, k)) {
    m.forEach((k, v) => fn(v, k));
  }
  static Map merge(Map a, Map b) {
    var m = new Map.from(a);
    if (b != null) {
      b.forEach((k, v) => m[k] = v);
    }
    return m;
  }
  static List<String> keys(Map<String, dynamic> a) {
    return a.keys.toList();
  }
  static bool isEmpty(Map m) => m.isEmpty;
  static bool equals(Map m1, Map m2) {
    if (m1.length != m2.length) {
      return false;
    }
    for (var key in m1.keys) {
      if (m1[key] != m2[key]) {
        return false;
      }
    }
    return true;
  }
}

class ListWrapper {
  static List clone(Iterable l) => new List.from(l);
  static List create() => new List();
  static List createFixedSize(int size) => new List(size);
  static get(List m, int k) => m[k];
  static void set(List m, int k, v) {
    m[k] = v;
  }
  static bool contains(List m, k) => m.contains(k);
  static List map(list, fn(item)) => list.map(fn).toList();
  static List filter(List list, bool fn(item)) => list.where(fn).toList();
  static int indexOf(List list, value, [int startIndex = 0]) =>
      list.indexOf(value, startIndex);
  static int lastIndexOf(List list, value, [int startIndex = null]) =>
      list.lastIndexOf(value, startIndex == null ? list.length : startIndex);
  static find(List list, bool fn(item)) =>
      list.firstWhere(fn, orElse: () => null);
  static bool any(List list, bool fn(item)) => list.any(fn);
  static void forEach(Iterable list, fn(item)) {
    list.forEach(fn);
  }
  static reduce(List list, fn(a, b), init) {
    return list.fold(init, fn);
  }
  static first(List list) => list.isEmpty ? null : list.first;
  static last(List list) => list.isEmpty ? null : list.last;
  static List reversed(List list) => list.reversed.toList();
  static void push(List l, e) {
    l.add(e);
  }
  static List concat(List a, List b) {
    return new List()
      ..length = a.length + b.length
      ..setRange(0, a.length, a)
      ..setRange(a.length, a.length + b.length, b);
  }
  static bool isList(l) => l is List;
  static void insert(List l, int index, value) {
    l.insert(index, value);
  }
  static removeAt(List l, int index) => l.removeAt(index);
  static void removeAll(List list, List items) {
    for (var i = 0; i < items.length; ++i) {
      list.remove(items[i]);
    }
  }
  static removeLast(List list) => list.removeLast();
  static bool remove(List list, item) => list.remove(item);
  static void clear(List l) {
    l.clear();
  }
  static String join(List l, String s) => l.join(s);
  static bool isEmpty(Iterable list) => list.isEmpty;
  static void fill(List l, value, [int start = 0, int end]) {
    l.fillRange(_startOffset(l, start), _endOffset(l, end), value);
  }
  static bool equals(List a, List b) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; ++i) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }
  static List slice(List l, [int from = 0, int to]) {
    return l.sublist(_startOffset(l, from), _endOffset(l, to));
  }
  static List splice(List l, int from, int length) {
    from = _startOffset(l, from);
    var to = from + length;
    var sub = l.sublist(from, to);
    l.removeRange(from, to);
    return sub;
  }
  static void sort(List l, [compareFn(a, b) = null]) {
    if (compareFn == null) {
      l.sort();
    } else {
      l.sort(compareFn);
    }
  }

  // JS splice, slice, fill functions can take start < 0 which indicates a position relative to
  // the end of the list
  static int _startOffset(List l, int start) {
    int len = l.length;
    return start = start < 0 ? max(len + start, 0) : min(start, len);
  }

  // JS splice, slice, fill functions can take end < 0 which indicates a position relative to
  // the end of the list
  static int _endOffset(List l, int end) {
    int len = l.length;
    if (end == null) return len;
    return end < 0 ? max(len + end, 0) : min(end, len);
  }
}

bool isListLikeIterable(obj) => obj is Iterable;

void iterateListLike(iter, fn(item)) {
  assert(iter is Iterable);
  for (var item in iter) {
    fn(item);
  }
}

class SetWrapper {
  static Set createFromList(List l) => new Set.from(l);
  static bool has(Set s, key) => s.contains(key);
}

File: modules/angular2/src/facade/collection.ts
import {isJsObject, global, isPresent} from 'angular2/src/facade/lang';

export var List = global.Array;
export var Map = global.Map;
export var Set = global.Set;
export var StringMap = global.Object;

// Safari and Internet Explorer do not support the iterable parameter to the
// Map constructor.  We work around that by manually adding the items.
var createMapFromPairs: {(pairs: List<any>): Map<any, any>} = (function() {
  try {
    if (new Map([1, 2]).size === 2) {
      return function createMapFromPairs(pairs: List<any>): Map<any, any> {
        return new Map(pairs);
      };
    }
  } catch (e) {
  }
  return function createMapAndPopulateFromPairs(pairs: List<any>): Map<any, any> {
    var map = new Map();
    for (var i = 0; i < pairs.length; i++) {
      var pair = pairs[i];
      map.set(pair[0], pair[1]);
    }
    return map;
  }
})();
var createMapFromMap: {(m: Map<any, any>): Map<any, any>} = (function() {
  try {
    if (new Map(new Map())) {
      return function createMapFromMap(m: Map<any, any>): Map<any, any> { return new Map(m); };
    }
  } catch (e) {
  }
  return function createMapAndPopulateFromMap(m: Map<any, any>): Map<any, any> {
    var map = new Map();
    m.forEach((v, k) => { map.set(k, v); });
    return map;
  }
})();
var _clearValues: {(m: Map<any, any>)} = (function() {
  if ((<any>(new Map()).keys()).next) {
    return function _clearValues(m: Map<any, any>) {
      var keyIterator = m.keys();
      var k;
      while (!((k = (<any>keyIterator).next()).done)) {
        m.set(k.value, null);
      }
    };
  } else {
    return function _clearValuesWithForeEach(m: Map<any, any>) {
      m.forEach((v, k) => { m.set(k, null); });
    }
  }
})();

export class MapWrapper {
  static create(): Map<any, any> { return new Map(); }
  static clone<K, V>(m: Map<K, V>): Map<K, V> { return createMapFromMap(m); }
  static createFromStringMap(stringMap): Map<string, any> {
    var result = MapWrapper.create();
    for (var prop in stringMap) {
      MapWrapper.set(result, prop, stringMap[prop]);
    }
    return result;
  }
  static createFromPairs(pairs: List<any>): Map<any, any> { return createMapFromPairs(pairs); }
  static get<K, V>(m: Map<K, V>, k: K): V { return m.get(k); }
  static set<K, V>(m: Map<K, V>, k: K, v: V) { m.set(k, v); }
  static contains<K>(m: Map<K, any>, k: K) { return m.has(k); }
  static forEach<K, V>(m: Map<K, V>, fn: /*(V, K) => void*/ Function) { m.forEach(<any>fn); }
  static size(m: Map<any, any>) { return m.size; }
  static delete<K>(m: Map<K, any>, k: K) { m.delete(k); }
  static clear(m: Map<any, any>) { m.clear(); }
  static clearValues(m: Map<any, any>) { _clearValues(m); }
  static iterable(m) { return m; }
  static keys<K>(m: Map<K, any>): List<K> { return m.keys(); }
  static values<V>(m: Map<any, V>): List<V> { return m.values(); }
}

/**
 * Wraps Javascript Objects
 */
export class StringMapWrapper {
  static create(): StringMap<any, any> {
    // Note: We are not using Object.create(null) here due to
    // performance!
    // http://jsperf.com/ng2-object-create-null
    return {};
  }
  static contains(map: StringMap<string, any>, key: string) { return map.hasOwnProperty(key); }
  static get<V>(map: StringMap<string, V>, key: string): V {
    return map.hasOwnProperty(key) ? map[key] : undefined;
  }
  static set<V>(map: StringMap<string, V>, key: string, value: V) { map[key] = value; }
  static keys(map: StringMap<string, any>): List<string> { return Object.keys(map); }
  static isEmpty(map: StringMap<string, any>) {
    for (var prop in map) {
      return false;
    }
    return true;
  }
  static delete (map: StringMap<string, any>, key: string) { delete map[key]; }
  static forEach<K, V>(map: StringMap<string, V>, callback: /*(V, K) => void*/ Function) {
    for (var prop in map) {
      if (map.hasOwnProperty(prop)) {
        callback(map[prop], prop);
      }
    }
  }

  static merge<V>(m1: StringMap<string, V>, m2: StringMap<string, V>): StringMap<string, V> {
    var m = {};

    for (var attr in m1) {
      if (m1.hasOwnProperty(attr)) {
        m[attr] = m1[attr];
      }
    }

    for (var attr in m2) {
      if (m2.hasOwnProperty(attr)) {
        m[attr] = m2[attr];
      }
    }

    return m;
  }

  static equals<V>(m1: StringMap<string, V>, m2: StringMap<string, V>): boolean {
    var k1 = Object.keys(m1);
    var k2 = Object.keys(m2);
    if (k1.length != k2.length) {
      return false;
    }
    var key;
    for (var i = 0; i < k1.length; i++) {
      key = k1[i];
      if (m1[key] !== m2[key]) {
        return false;
      }
    }
    return true;
  }
}

export class ListWrapper {
  static create(): List<any> { return new List(); }
  static createFixedSize(size): List<any> { return new List(size); }
  static get(m, k) { return m[k]; }
  static set(m, k, v) { m[k] = v; }
  static clone(array: List<any>) { return array.slice(0); }
  static map(array, fn) { return array.map(fn); }
  static forEach(array: List<any>, fn: Function) {
    for (var i = 0; i < array.length; i++) {
      fn(array[i]);
    }
  }
  static push(array, el) { array.push(el); }
  static first(array) {
    if (!array) return null;
    return array[0];
  }
  static last(array) {
    if (!array || array.length == 0) return null;
    return array[array.length - 1];
  }
  static find(list: List<any>, pred: Function) {
    for (var i = 0; i < list.length; ++i) {
      if (pred(list[i])) return list[i];
    }
    return null;
  }
  static indexOf(array: List<any>, value, startIndex = 0) {
    return array.indexOf(value, startIndex);
  }
  static reduce<T, E>(list: List<T>,
                      fn: (accumValue: E, currentValue: T, currentIndex: number, array: T[]) => E,
                      init: E) {
    return list.reduce(fn, init);
  }
  static filter(array, pred: Function) { return array.filter(pred); }
  static any(list: List<any>, pred: Function) {
    for (var i = 0; i < list.length; ++i) {
      if (pred(list[i])) return true;
    }
    return false;
  }
  static contains(list: List<any>, el) { return list.indexOf(el) !== -1; }
  static reversed(array) {
    var a = ListWrapper.clone(array);
    return a.reverse();
  }
  static concat(a, b) { return a.concat(b); }
  static isList(list) { return Array.isArray(list); }
  static insert(list, index: int, value) { list.splice(index, 0, value); }
  static removeAt(list, index: int) {
    var res = list[index];
    list.splice(index, 1);
    return res;
  }
  static removeAll(list, items) {
    for (var i = 0; i < items.length; ++i) {
      var index = list.indexOf(items[i]);
      list.splice(index, 1);
    }
  }
  static removeLast<T>(list: List<T>): T { return list.pop(); }
  static remove(list, el): boolean {
    var index = list.indexOf(el);
    if (index > -1) {
      list.splice(index, 1);
      return true;
    }
    return false;
  }
  static clear(list) { list.splice(0, list.length); }
  static join(list, s) { return list.join(s); }
  static isEmpty(list) { return list.length == 0; }
  static fill(list: List<any>, value, start: int = 0, end: int = null) {
    list.fill(value, start, end === null ? undefined : end);
  }
  static equals(a: List<any>, b: List<any>): boolean {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; ++i) {
      if (a[i] !== b[i]) return false;
    }
    return true;
  }
  static slice<T>(l: List<T>, from: int = 0, to: int = null): List<T> {
    return l.slice(from, to === null ? undefined : to);
  }
  static splice<T>(l: List<T>, from: int, length: int): List<T> { return l.splice(from, length); }
  static sort<T>(l: List<T>, compareFn?: (a: T, b: T) => number) {
    if (isPresent(compareFn)) {
      l.sort(compareFn);
    } else {
      l.sort();
    }
  }
  static toString<T>(l: List<T>): string { return l.toString(); }
}

export function isListLikeIterable(obj): boolean {
  if (!isJsObject(obj)) return false;
  return ListWrapper.isList(obj) ||
         (!(obj instanceof Map) &&  // JS Map are iterables but return entries as [k, v]
          Symbol.iterator in obj);  // JS Iterable have a Symbol.iterator prop
}

export function iterateListLike(obj, fn: Function) {
  if (ListWrapper.isList(obj)) {
    for (var i = 0; i < obj.length; i++) {
      fn(obj[i]);
    }
  } else {
    var iterator = obj[Symbol.iterator]();
    var item;
    while (!((item = iterator.next()).done)) {
      fn(item.value);
    }
  }
}


// Safari and Internet Explorer do not support the iterable parameter to the
// Set constructor.  We work around that by manually adding the items.
var createSetFromList: {(lst: List<any>): Set<any>} = (function() {
  var test = new Set([1, 2, 3]);
  if (test.size === 3) {
    return function createSetFromList(lst: List<any>): Set<any> { return new Set(lst); };
  } else {
    return function createSetAndPopulateFromList(lst: List<any>): Set<any> {
      var res = new Set(lst);
      if (res.size !== lst.length) {
        for (var i = 0; i < lst.length; i++) {
          res.add(lst[i]);
        }
      }
      return res;
    };
  }
})();
export class SetWrapper {
  static createFromList<T>(lst: List<T>): Set<T> { return createSetFromList(lst); }
  static has<T>(s: Set<T>, key: T): boolean { return s.has(key); }
}
